File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        priority_queue<int> pq; //Max Heap 
        int cB = 0,lad=ladders,n = heights.size();
        for(int i=1;i<n;i++)
        {
            if(heights[i]>heights[i-1])
            {
                int diff = heights[i]-heights[i-1];
                cB += diff; //Current No. Of Bricks Required
                pq.push(diff);
                if(cB>bricks) //If required no. of bricks exceeds the 
				{				//available bricks
                    if(lad>0)
                    {
						//Since we have ladder left, we can either directly
						//use it here or we can imagine that we can travel back
						//to the part where we have used the maximum no. of bricks
						//use the ladder there and take back those bricks such
						//that we can use them here.
                        cB-=pq.top();
                        pq.pop();
                        lad--;
                    }
                    else
                    {
                        return i-1;
                    }
                }
            }
        }
		//We reached the last building
        return n-1;
    }
};

File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.java

import java.util.*;

class Solution {
	public static int furthestBuilding(int[] heights, int bricks, int ladders) {
		PriorityQueue<Integer> maxheap = new PriorityQueue<>(Collections.reverseOrder());
		int i = 1;
		for (i = 1; i < heights.length; i++) {
			int diff = heights[i] - heights[i - 1];
			if (diff <= 0)
				continue;

			bricks -= diff;
			maxheap.add(diff);

			if (bricks < 0) {
				bricks += maxheap.poll();
				ladders--;
			}

			if (ladders < 0)
				break;
		}

		return --i;
	}

	public static void check(int actual, int expected, int tc) {
		System.out.print("TEST CASE " + tc + "  - ");
		System.out.println(expected == actual ? "PASSED" : "FAILED\nExpected: " + expected + "\nActual: " + actual);
	}

	public static void main(String[] args) {
		check(furthestBuilding(new int[] { 4, 2, 7, 6, 9, 14, 12 }, 5, 1), 4, 1);
		check(furthestBuilding(new int[] { 4, 12, 2, 7, 3, 18, 20, 3, 19 }, 10, 2), 7, 2);
		check(furthestBuilding(new int[] { 14, 3, 19, 3 }, 17, 0), 3, 3);
	}
}